### PR TITLE
Weekly pattern text summarization

### PR DESCRIPTION
**Pull request by jcc37**
Add `TrendWeeklyPattern` summarization module and tests for time-series data.
The summaries describe the common (average) weekly pattern in the time-series data. The following are sample summaries for a one-month time-series data:
* The active users from Monday to Friday increased by 16.5 per day on average.
* The active users from Friday to Sunday decreased by 63.0 per day on average.

Several time-series util functions and trend core functions are also added or updated to support this summarization module.

### CODE REVIEW

File: src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    const otherSummaries = summaries.filter(isOtherSummary);
+
+    expect(mondayToSundaySummary.length).toBe(1);
+    expect(mondayToSundaySummary.every(hasHighValidity)).toBeTrue();
+    expect(otherSummaries.every(hasLowValidity)).toBeTrue();
+  });
+
+  it('should create summaries describing constant data.', () => {
+    const points: TimeSeriesPoint[] = [];
+    for (let i = 6; i <= 26; i++) {
```
Comment (Rhombus33) [inline src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts:68]: add a note that the test data starts on Saturday and why.

File: src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    const otherSummaries = summaries.filter(isOtherSummary);
+
+    expect(mondayToWednesdaySummary.length).toBe(1);
+    expect(mondayToWednesdaySummary.every(hasHighValidity)).toBeTrue();
+    expect(wednesdayToSundaySummary.length).toBe(1);
+    expect(wednesdayToSundaySummary.every(hasHighValidity)).toBeTrue();
+    expect(otherSummaries.every(hasLowValidity)).toBeTrue();
+  });
+
+  it('should create summaries describing V shape weekly pattern in linearly increasing data.', () => {
```
Comment (Rhombus33) [inline src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts:116]: "linearly increasing data". Does this mean linearly increase between the same day of the week (i.e. week1 Monday -> week2 Monday -> week3 Monday)

Not easy to understand what the test case is here and for the tests below with "linearly increasing" or "decreasing" data.

**Review by Rhombus33 — APPROVED** [review]

File: src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts (hunk 1 above)
Comment (jcc37) [reply src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts:68]: No it starts on Monday. `new Date(2020, 6, 6)` is a Monday

Comment (jcc37) [reply src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts:68]: No it starts on Monday. `new Date(2020, 6, 6)` is a Monday

File: src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts (hunk 2 above)
Comment (jcc37) [reply src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts:116]: This is an example of "weekly pattern in linearly increasing data". Basically it is "linear increasing data + pattern offset for the day of the week":
image

File: src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts (hunk 1 above)
Comment (Rhombus33) [reply src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts:68]: Then you should add a note that it starts on Monday and why :P

Comment (jcc37) [reply src/datasets/summarizations/trend-weekly-pattern.summarization.spec.ts:68]: added in #107

**Pull request merged**